qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,217 @@
+    """
+    Get a tuple of required/nonrequired inputs or outputs for each method
+
+    Parameters
+    ----------
+    action : Qiime2.action
+
+    type : {0,1,2}
+        Delineates if getting the action input, param, or output types
+
+    Returns
+    -------
+    Tuple - list of required and optional semantic types
+
+    """
+
+    req = []
+    non_req = []
+
+    if type == 0:
+        for k, v in action.signature.inputs.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+    elif type == 1:
+        for k, v in action.signature.parameters.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+    else:
+        for k, v in action.signature.outputs.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+
+    return req, non_req
+
+
+def get_combinations(no_req):
+    """
+    Get all permutations of list of semantic types
+
+    Parameters
+    ----------
+    non_req : list of semantic types
+
+    Returns
+    -------
+    List of lists - each list contains an instance of a permutation
+
+    """
+
+    if not no_req:
+        return no_req
+    no_req_comb = []
+    for i in range(len(no_req)+1):
+        combs = combinations(no_req, i)
+        for c in combs:
+            no_req_comb.append(list(c))
+    return no_req_comb
+
+
+def generate_nodes_by_action(action):
+    """
+    Given a method, generates all combinations of inputs and
+    outputs for that particular method and and stores the combinations
+    as dictionaries in a resulting list.
+
+    Parameters
+    ----------
+    method : Qiime2.action
+
+    Returns
+    -------
+    list of dictionaries - each dictionary is a combination inputs and
+        outputs for particular node
+
+    """
+
+    results = []
+
+    input, input_nr = get_next_arguments(action, 0)
+    param, param_nr = get_next_arguments(action, 1)
+    output, _ = get_next_arguments(action, 2)
+
+    non_req = []
+
+    # unravel potential unions
+    input = unravel_union(input)
+    param = unravel_union(param)
+
+    for x in input_nr + param_nr:
+        non_req += unravel_union(x)
+
+    for i in input:
+        for p in param:
+            if non_req:
+                non_req = get_combinations(non_req)
+                for nr in non_req:
+                    d = {}
+                    d['inputs'] = i
+                    d['params'] = p
+                    d['outputs'] = output
+                    d['non_req'] = nr
+                    results.append(d)
+            else:
+                d = {}
+                d['inputs'] = i
+                d['params'] = p
+                d['outputs'] = output
+                d['non_req'] = []
+                results.append(d)
```

Comment:
**Discussion**
I think you could simplify this entire loop if the following was done:

1. Create a function `unravel` such that for every parameter (cs-definition of the term) of an actions signature such that:
   - when it encounters an optional value, it yields None and then yields the value
   - when it encounters a union, it unravels that (and if it is optional it also yields None).

2. zip or otherwise combine the unraveled iterable with the parameter name, so that you now have an iterable of `(name, type)` for each parameter

3. use `itertools.product` on *all* iterables, this will result in a new iterable of tuples where the tuple contains a `(name, type)` value for each parameter from the step above.

4. For each tuple-of-tuple convert that to a dict (you can actually just use `dict(tuple-of-tuple)` to construct it)

5. yield that dict